salt/utils/path.py

```diff
@@ -412,3 +417,94 @@ def os_walk(top, *args, **kwargs):
         top_query = salt.utils.stringutils.to_str(top)
     for item in os.walk(top_query, *args, **kwargs):
         yield salt.utils.data.decode(item, preserve_tuples=True)
+
+
+def prepend_root_dir(path, root_dir=None):
+    '''
+    Prepends 'path' with value of the 'root_dir' option.
+    '''
+    if root_dir:
+        root_dir = os.path.abspath(root_dir)
+
+    def_root_dir = salt.syspaths.ROOT_DIR.rstrip(os.sep)
+    tmp_path_def_root_dir = None
+    tmp_path_root_dir = None
+
+    # When running testsuite, salt.syspaths.ROOT_DIR is often empty
+    if path == def_root_dir or path.startswith(def_root_dir + os.sep):
+        # Remove the default root dir prefix
+        tmp_path_def_root_dir = path[len(def_root_dir):]
+
+    if root_dir and (path == root_dir or
+                     path.startswith(root_dir + os.sep)):
+        # Remove the root dir prefix
+        tmp_path_root_dir = path[len(root_dir):]
+
+    if tmp_path_def_root_dir and not tmp_path_root_dir:
+        # Just the default root dir matched
+        path = tmp_path_def_root_dir
+    elif tmp_path_root_dir and not tmp_path_def_root_dir:
+        # Just the root dir matched
+        path = tmp_path_root_dir
+    elif tmp_path_def_root_dir and tmp_path_root_dir:
+        # In this case both the default root dir and the override root
+        # dir matched; this means that either
+        # def_root_dir is a substring of root_dir or vice versa
+        # We must choose the most specific path
+        if def_root_dir in root_dir:
+            path = tmp_path_root_dir
+        else:
+            path = tmp_path_def_root_dir
+    elif salt.utils.platform.is_windows() and not os.path.splitdrive(path)[0]:
+        # In windows, os.path.isabs resolves '/' to 'C:\\' or whatever
+        # the root drive is.  This elif prevents the next from being
+        # hit, so that the root_dir is prefixed in cases where the
+        # drive is not prefixed on a config option
+        pass
+    elif os.path.isabs(path):
+        # Absolute path (not default or overridden root_dir)
+        # No prepending required
+        return path
+
+    if root_dir is None:
+        # Transform to absolute path without prepending root_dir
+        return os.path.abspath(path)
+
+    # Prepending the root dir
+    return os.path.join(root_dir, path)
+
+
+def expand_path(path, root_dir=None):
+    '''
+    Resolves user home '~' or prepend the 'root_dir'
+    directory and returns the expanded 'path'
+    '''
+    if path.startswith('~'):
+        path = os.path.expanduser(path)
+    elif urlparse(path).scheme == '':
```

Comment:
```python
elif not urlparse(path).scheme:
```

Reply:
That was a cut&past from `salt.config`, there are several of them current `develop` branch:

```
moi@home:~/src/salt (develop)$ ack -l -c -Q "scheme == ''" 
salt/config/__init__.py:5
salt/modules/mac_pkgutil.py:1
```

How should we handle the other ones?

Reply:
We will get there one day. :wink: Not everything needs to be copypasted. Just take care of your own stuff it is great and make it pythonic enough. The rest will be "combed" soon.

Reply:
Sure, but a task like this can be [tagged](https://docs.saltstack.com/en/latest/topics/development/labels.html#labels-and-milestones) as `Minor change` or `Medium Change` and something like `first good issue` [as I saw somewhere else](https://perl6.party/post/Introducing-Newcomer-Guide-to-Contributing-to-Core-Perl-6) to help newcomers.